### PR TITLE
Cleanup of grpc package

### PR DESCRIPTION
**Pull request by peter-edge**
This is a cleanup of the grpc package, partially in preparation to handle multiple encodings per endpoint. Two main things are done:

- `Transport` is propagated to the `Inbound` and `Outbound` structs, and the configuration code is updated accordingly. This is done to match other packages, and to implement `Transports` on both `Inbound` and `Outbound`.
- `handler` structs no longer take a yarpc service name or encoding, but are generic passthroughs to `Router.Choose` that construct `transport.Request` structs. This removes setting a default encoding per handler. The code to create each of these handlers is updated as well. This ends up making the code significantly cleaner IMO.

### CODE REVIEW

**Comment by kriskowal** [comment]
Needs a rebase. Looking at commits for now.

File: transport/x/grpc/inbound.go [hunk 1]
```diff
@@ -134,45 +135,42 @@ func (i *Inbound) getServiceDescs() ([]*grpc.ServiceDesc, error) {
 	if len(procedures) == 0 {
 		return nil, errRouterHasNoProcedures
 	}
-	grpcServiceNameToServiceDesc := make(map[string]*grpc.ServiceDesc)
+
+	serviceNameToMethodNames := make(map[string]map[string]bool)
```
Comment (breerly) [inline transport/x/grpc/inbound.go]: aside - consider factoring this block out into its own func so this method is more readable

File: transport/x/grpc/inbound.go [hunk 2]
```diff
@@ -134,45 +135,42 @@ func (i *Inbound) getServiceDescs() ([]*grpc.ServiceDesc, error) {
...
+			methodNames = make(map[string]bool)
+			serviceNameToMethodNames[serviceName] = methodNames
 		}
-		serviceDesc.Methods = append(serviceDesc.Methods, methodDesc)
+		methodNames[methodName] = true
 	}
-	serviceDescs := make([]*grpc.ServiceDesc, 0, len(grpcServiceNameToServiceDesc))
-	for _, serviceDesc := range grpcServiceNameToServiceDesc {
+
+	serviceDescs := make([]*grpc.ServiceDesc, 0, len(serviceNameToMethodNames))
```
Comment (breerly) [inline transport/x/grpc/inbound.go]: aside - looks like this big block might also be more neatly tucked away into its own func

File: transport/x/grpc/inbound.go (hunk 1 above)
Comment (peter-edge) [inline transport/x/grpc/inbound.go]: it's harder to do this with this block, i think this is fine

File: transport/x/grpc/inbound.go (hunk 2 above)
Comment (peter-edge) [inline transport/x/grpc/inbound.go]: i think this is actually the simplest route

File: transport/x/grpc/inbound.go [hunk 3]
```diff
@@ -42,14 +42,15 @@ var (
...
+	t              *Transport
 	listener       net.Listener
 	inboundOptions *inboundOptions
 	router         transport.Router
 	server         *grpc.Server
 }
 
-func newInbound(listener net.Listener, options ...InboundOption) *Inbound {
-	return &Inbound{internalsync.Once(), sync.Mutex{}, listener, newInboundOptions(options), nil, nil}
+func newInbound(t *Transport, listener net.Listener, options ...InboundOption) *Inbound {
```
Comment (willhug) [inline transport/x/grpc/inbound.go]: should this be a function of the transport?

File: transport/x/grpc/outbound.go [hunk 4]
```diff
@@ -47,13 +47,14 @@ var _ transport.UnaryOutbound = (*Outbound)(nil)
...
 	lock            sync.Mutex
+	t               *Transport
 	address         string
 	outboundOptions *outboundOptions
 	clientConn      *grpc.ClientConn
 }
 
-func newSingleOutbound(address string, options ...OutboundOption) *Outbound {
-	return &Outbound{internalsync.Once(), sync.Mutex{}, address, newOutboundOptions(options), nil}
+func newSingleOutbound(t *Transport, address string, options ...OutboundOption) *Outbound {
```
Comment (willhug) [inline transport/x/grpc/outbound.go]: ditto on being a function on the transport.

Comment (willhug) [inline transport/x/grpc/outbound.go]: just saw the other functions, all private, sounds good.

**Pull request merged**